feat(handler): Allow specifying subsystem and category in Handler

jeff-foster-trabon opened the pull request:
The Handler class now accepts optional subsystem and category arguments in its constructor, to allow setting them from `logging.config.dictConfig()`

jeff-foster-trabon committed:
feat(handler): Allow specifying subsystem and category in Handler

The Handler class now accepts optional subsystem and category arguments in its constructor, to allow setting them from `logging.config.dictConfig()`

simonrob commented on pyoslog/handler.py:
```diff
@@ -10,10 +11,28 @@ class Handler(logging.Handler):
+                subsystem, category if category is not None else 'default'
+            )
+        else:
+            self._log_object = OS_LOG_DEFAULT
```
I think it's better to use `setSubsystem` here – e.g.:

```python
        if subsystem is not None:
            self.setSubsystem(subsystem, category)
        else:
            self._log_object = OS_LOG_DEFAULT
```

simonrob commented on pyoslog/handler.py:
```diff
@@ -10,10 +11,28 @@ class Handler(logging.Handler):
+    def __init__(
+        self, subsystem: Optional[str] = None, category: Optional[str] = None
+    ) -> None:
+        """Initialise a Handler instance.
```
The note about output type got lost – I'd add something like `logging at OS_LOG_TYPE_DEFAULT`.

simonrob commented on pyoslog/core.py:
```diff
@@ -24,7 +24,12 @@ def _get_os_log_default():
+        log_object: _os_log_t_native_type,
+        subsystem: Optional[str],
+        category: Optional[str],
+    ) -> None:
```
The `subsystem` and `category` values are not actually meant to be optional on the user side. Setting these to `None` is how the `OS_LOG_DISABLED` and (including `log_object`) `OS_LOG_DEFAULT` objects are created, but that is a special case.

simonrob reviewed (commented):
Thanks for the PR!

I've added a few comments. Please could you also look at suitable tests? If you make my recommended change to just call `setSubsystem`, tests could be simply checking the created object in a  new custom `Handler` around [here](https://github.com/simonrob/pyoslog/blob/6546824b58f04687c9dc4dfe6349ed096b1dda6d/tests/test_handler.py#L33C1-L35C61).

jeff-foster-trabon committed:
feat(handler): Address review comments

Refactors `Handler.__init__` to explicitly use `setSubsystem` when a subsystem is provided.
Revert typing changes to the `os_log_t` constructor signature.
Adds new tests for Handler initializations with subsystem and category.

jeff-foster-trabon commented:
fwiw I struggled a bit getting the tests running. There was always an xpc entry after the pyoslog one, so `get_latest_log_message` returned that. I ran the tests by temporarily changing it to `return recent_messages[-2]`.

Maybe a difference in overall log system config? I'm running Sonoma 14.7.5

simonrob commented:
Thanks!

I'll take a look at the test setup – it's always been a challenge to reliably read the message during testing given all of the external configuration aspects that can affect it, but hopefully I can improve this a little.

simonrob commented:
I looked at this, and ultimately decided to just strip out the XPC messages (see 926d7fe). Could you check whether this also fixes things for you?

jeff-foster-trabon committed:
test: make 'python' process name check case-insensitive

jeff-foster-trabon commented:
Yes now the build script with tests ran for me with no issues—after making a couple of the checks for process name _python_ case-insensitive, as the process name is _Python_ in my setup.

I rebased onto your changes and force pushed

simonrob committed:
Minor reformatting to match code style
